pkg/sync/sync_file.go

```diff
@@ -577,6 +581,275 @@ func (sf *SyncingFile) waitForRestoreComplete() (err error) {
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to prepare target file")
+	}
+
+	sourceFileReader, sourceLoopDevicePath, err := sf.openCloneSourceFile(sourceFile, encryption, credential)
+	defer func() {
+		if sourceFileReader != nil {
+			sourceFileReader.Close()
+		}
+		if sourceLoopDevicePath == "" {
+			sf.closeCryptoDevice(sourceLoopDevicePath)
+		}
+	}()
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to open clone source file")
+	}
+
+	targetFileWriter, targetLoopDevicePath, err := sf.openCloneTargetFile(encryption, credential)
+	defer func() {
+		if targetFileWriter != nil {
+			targetFileWriter.Close()
+		}
+		if targetLoopDevicePath == "" {
+			sf.closeCryptoDevice(targetLoopDevicePath)
+		}
+	}()
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to open clone source file")
+	}
+
+	nw, err := IdleTimeoutCopy(sf.ctx, sf.cancel, sourceFileReader, targetFileWriter, sf, writeZero)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to copy the data with timeout")
+	}
+	return nw, err
+}
+
+func (sf *SyncingFile) openCloneSourceFile(sourceFile string, encryption types.EncryptionType, credential map[string]string) (*os.File, string, error) {
+	loopDevicePath := ""
+	if encryption == types.EncryptionTypeDecrypt {
+		loopDevicePath, err := sf.setupCryptoDevice(sourceFile, false, credential)
+		if err != nil {
+			return nil, loopDevicePath, errors.Wrapf(err, "failed to setup the crypto device with the file %v during cloning", sourceFile)
+		}
+		sourceFile = types.BackingImageMapper(sf.uuid)
+	}
+
+	if _, err := os.Stat(sourceFile); err != nil {
+		return nil, loopDevicePath, errors.Wrapf(err, "%v not found during cloning", sourceFile)
+	}
+
+	sourceFileReader, err := os.Open(sourceFile)
+	if err != nil {
+		return nil, loopDevicePath, errors.Wrapf(err, "Error opening file %v", sourceFile)
+	}
+
+	return sourceFileReader, loopDevicePath, nil
+}
+
+func (sf *SyncingFile) openCloneTargetFile(encryption types.EncryptionType, credential map[string]string) (*os.File, string, error) {
+	loopDevicePath := ""
+	targetFile := sf.tmpFilePath
+
+	if encryption == types.EncryptionTypeEncrypt {
+		loopDevicePath, err := sf.setupCryptoDevice(targetFile, true, credential)
+		if err != nil {
+			return nil, loopDevicePath, errors.Wrapf(err, "failed to setup the crypto device with the file %v during cloning", targetFile)
+		}
+		targetFile = types.BackingImageMapper(sf.uuid)
+	}
+
+	if _, err := os.Stat(targetFile); err != nil {
+		return nil, loopDevicePath, errors.Wrapf(err, "%v not found during cloning", targetFile)
+	}
+
+	targetFileWriter, err := os.OpenFile(targetFile, os.O_RDWR, 0666)
+	if err != nil {
+		return nil, loopDevicePath, errors.Wrapf(err, "Error opening file %v", targetFile)
+	}
+	return targetFileWriter, loopDevicePath, nil
+}
+
+func (sf *SyncingFile) prepareCloneSourceFile(sourceBackingImage, sourceBackingImageUUID string, encryption types.EncryptionType) (string, string, bool, error) {
```

Comment:
Need to check if the source file exists (is on the same node as the target file) in this function.

Reply:
fixed, thanks